<!DOCTYPE html>
<html>
<head>
<title>README.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<ul>
<li>TOC
{:toc}</li>
</ul>
<h1 id="detecting-horizontal-transfer-of-transposons-htt">Detecting Horizontal Transfer of Transposons (HTT)</h1>
<p>{:.no_toc}</p>
<h2 id="introduction">Introduction</h2>
<p>The specific steps for the identification of HTT events are typically
applied to two scenarios with distinct workflows (A and B in Figure 1)
that converge at the stage where HTT candidates are validated (C in
Figure 1). Workflows A, B and C are described below, with numbered steps
in each section that link back to Figure 1.</p>
<p><img src="HTT_detection_workflow.png" alt="Figure 1, The HTT detection flowchart"></p>
<h2 id="prerequisites">Prerequisites</h2>
<h3 id="recommended-programstools-a-name%22recommended-programs%22">Recommended programs/tools <a name="recommended-programs" /></h3>
<ul>
<li>BLAST (<a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">https://blast.ncbi.nlm.nih.gov/Blast.cgi</a>)</li>
<li>SAMtools (<a href="http://www.htslib.org/">http://www.htslib.org/</a>)</li>
<li>BEDtools (<a href="http://bedtools.readthedocs.io/en/latest/">http://bedtools.readthedocs.io/en/latest/</a>)</li>
<li>CENSOR, which requires wu-blast and bioperl
(<a href="https://girinst.org/downloads/software/censor/">https://girinst.org/downloads/software/censor/</a>)</li>
<li>USEARCH (<a href="https://www.drive5.com/usearch/">https://www.drive5.com/usearch/</a>)</li>
<li>VSEARCH (<a href="https://github.com/torognes/vsearch">https://github.com/torognes/vsearch</a>)</li>
<li>MUSCLE (<a href="https://www.drive5.com/muscle/">https://www.drive5.com/muscle/</a>)</li>
<li>MAFFT (<a href="https://mafft.cbrc.jp/alignment/software/">https://mafft.cbrc.jp/alignment/software/</a>)</li>
<li>RepeatMasker (<a href="http://www.repeatmasker.org/">http://www.repeatmasker.org/</a>)</li>
<li>Repeatmodeler (<a href="https://www.repeatmasker.org/RepeatModeler/">https://www.repeatmasker.org/RepeatModeler/</a>)</li>
<li>Alignment viewer (e.g. JalView, <a href="https://www.jalview.org/">https://www.jalview.org/</a>)</li>
<li>R (<a href="https://www.r-project.org/">https://www.r-project.org/</a>)
<ul>
<li>Required packages: tidyverse, plyranges, BSgenome, optranges</li>
</ul>
</li>
</ul>
<h3 id="optional-programstools-a-name%22optional-programs%22">Optional programs/tools <a name="optional-programs" /></h3>
<ul>
<li>CARP
(<a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0193588">https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0193588</a>)</li>
<li>LASTZ (https://www.bx.psu.edu/~rsharris/lastz/)</li>
<li>SiLiX (<a href="http://lbbe.univ-lyon1.fr/-SiLiX-?lang=en">http://lbbe.univ-lyon1.fr/-SiLiX-?lang=en</a>)</li>
<li>Gblocks
(<a href="https://ls23l.lscore.ucla.edu/MakeTree/documentation/gblocks.html">https://ls23l.lscore.ucla.edu/MakeTree/documentation/gblocks.html</a>)</li>
<li>HMMer (<a href="http://hmmer.org/">http://hmmer.org/</a>)</li>
<li>FastTree (<a href="http://www.microbesonline.org/fasttree/">http://www.microbesonline.org/fasttree/</a>)</li>
<li>IQ-TREE (<a href="http://www.iqtree.org/">http://www.iqtree.org/</a>)</li>
<li>CD-HIT (<a href="http://weizhong-lab.ucsd.edu/cd-hit/">http://weizhong-lab.ucsd.edu/cd-hit/</a>)</li>
</ul>
<h3 id="for-users-a-name%22for-users%22">For users <a name="for-users" /></h3>
<p>Our methods are implemented in Linux/R and we assume that users will be
familiar with the bash shell and R. Some level of familiarity with
queuing systems in HPC is also recommended.</p>
<h2 id="workflows">Workflows</h2>
<h3 id="a-ab-initio-workflow-a-name%22workflowa%22">A: Ab initio workflow <a name="workflowA" /></h3>
<p>Pipeline A is designed to identify potential horizontally transferred
sequences in newly sequenced genomes based on their absence from the
genome of a closely related species.</p>
<p>All input files should be FASTA files. Raw repeats can be from any <em>ab
initio</em> software package. Input genomes should be in
<code>Pipeline_A/genomes/</code> folder.</p>
<p><em><strong>Usage</strong></em></p>
<p>NOTE: Assume you are in <code>Pipeline_A</code> folder to run the following scripts.</p>
<p><strong>0) Ab initio TE annotation using RepeatModeler.</strong></p>
<p>Run <a href="./Pipeline_A/HT_stage_0.sh">HT_stage_0.sh</a></p>
<pre class="hljs"><code><div>GENOME=&lt;source_genome&gt; THREADS=&lt;number of threads to use&gt; bash HT_stage_0.sh 
</div></code></pre>
<p><strong>1) Cluster and perform initial sweep and generate multiple alignments
for manual curation of repeats.</strong></p>
<p>Run <a href="./Pipeline_A/HT_stage_1.sh">HT_stage_1.sh</a></p>
<pre class="hljs"><code><div>GENOME=&lt;source_genome&gt; OUTGROUP=&lt;outgroup_genome&gt; QUERY=&lt;file_containing_raw_repeats&gt; THREADS=&lt;number of threads to use&gt; bash HT_stage_1.sh
</div></code></pre>
<p>This script performs an initial search for sequences which have 2 or
more copies in the query genome and are absent from the outgroup genome.
If any sequences are absent from the outgroup a multiple sequence alignment will be
created for curation.</p>
<p><strong>2) Manually curate the alignments in Geneious, JalView or the like.</strong></p>
<p>This is necessary as most <em>ab initio</em> aligners do not capture full
repeats, and some families may be classified incorrectly. Additionally,
this step can reveal redundant sequences not removed from clustering,
e.g. non-autonomous DNA transposons derived from autonomous DNA
transposons.</p>
<p><strong>3) Confirm HTT condidates based on either repeating local alignment
finding absence or higher-than-expected-divergence from the genomes of
them most closely related species.</strong></p>
<p>Run <a href="./Pipeline_A/HT_stage_2.sh">HT_stage_2.sh</a> using curated
repeats as input.</p>
<pre class="hljs"><code><div>GENOME=&lt;source_genome&gt; SPECIES=&lt;name_of_source_species&gt; OUTGROUP=&lt;outgroup_genome&gt; QUERY=&lt;file_containing_curated_repeats&gt; THREADS=&lt;number of threads to use&gt; bash HT_stage_2.sh
</div></code></pre>
<p>This script carries out a validation of the initial search using
consensus sequences generated from the curation step. This is necessary
as fragmented repeats which appeared to be mssing from an outgroup
species may in fact be present. For example, when searching with TEs
from a seal genome using a mustelid as the outgroup, stage 1 identified
4 L1 fragments which appeared to be absent from the mink. After curation
it became clear those L1s were in fact present in the mustelid, just not
identifed in the initial sweep, likely due to their
truncation/fragmentation. The initial curation step fixes this problem
by ensuring that searches of the outgroup genome are done with queries
that contain the complete TE of interest.</p>
<p><strong>4) Local alignment of HTT condidates against all available genomes to
identify most closely related sequences and species with no hits.</strong></p>
<p>Run <a href="./Pipeline_A/HT_stage_3.sh">HT_stage_3.sh</a></p>
<pre class="hljs"><code><div>GENOME=&lt;name_of_source_species&gt; OUTGROUPS=&lt;file_containing_list_of_genomes&gt; QUERY=&lt;file_containing_curated_repeats&gt; THREADS=&lt;number of threads to use&gt; bash HT_stage_2.sh
</div></code></pre>
<p>This script searches for repeats verified as HTT candidates, searches for them in other species (from a list provided) and creates a MSA of each repeats if identified in a species ready for curation. This script is written to utilise genomes downloaded from Genbank and present in the <code>Pipeline_A/genomes/</code> folder and gzipped. If necessary modify this script for your particular situation.</p>
<p><em><strong>Example usage for Pipeline A</strong></em></p>
<ul>
<li>
<ol>
<li>run genome_downloader.sh to download all high quality snake and echinoderm genomes from GenBank (requires Entrez Direct)</li>
</ol>
</li>
<li>
<ol start="2">
<li>unzip <em>Laticauda</em> colubrina genome to use as source genome and <em>Naja naja</em> genome to use as outgroup</li>
</ol>
</li>
<li>
<ol start="3">
<li>run HT_stage_1.sh (example RepeatModeler output of <em>Laticauda colubrina</em> is in the data folder)</li>
</ol>
</li>
<li>
<ol start="4">
<li>manually curate potential HTT candidates</li>
</ol>
</li>
<li>
<ol start="5">
<li>run HT_stage_2.sh using curated HTT candidates</li>
</ol>
</li>
<li>
<ol start="6">
<li>run HT_stage_3.sh to examine other species in which HTT candidates are present in. With the example dataset several <em>Harbingers</em> identified in <em>Laticauda colubrina</em> will also be present in <em>Laticauda laticaudata</em> and various echinoderms</li>
</ol>
</li>
</ul>
<h3 id="b-workflow-for-global-htt-screening-of-specific-tes-a-name%22workflowb%22">B: Workflow for global HTT screening of specific TEs <a name="workflowB" /></h3>
<p>Pipeline B is used for detecting potential horizontal transfer events, starting from a set of curated repeat consensus sequences from available sources (e.g. RepBase or Dfam) for the TE of interest.</p>
<p>It is a simplified version of the code used to infer horizontal transfer events involving L1 and BovB retrotransposons in eukaryotes (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1456-7">Ivancevic et al., <em>Genome Biology</em>, 2018</a>).</p>
<p><em><strong>Usage</strong></em></p>
<p>NOTE: Assume you are in <code>Pipeline_B</code> to run following scripts.</p>
<p><strong>0) Download and prepare the genomes you want to screen.</strong></p>
<p><em>0a) Append species names to the genome name.</em></p>
<p>As the names given to genome assemblies are not usually informative, you will want to append species names to the genome names.</p>
<p>Run <a href="./Pipeline_B/0a_rename_genome.sh">0a_rename_genome.sh</a>.</p>
<p>Example usage:</p>
<pre class="hljs"><code><div>GENOME=&lt;source_genome&gt; SPECIES=&lt;species_name&gt; bash 0a_rename_genome.sh
</div></code></pre>
<p><em>0b) Make each genome a BLAST database and create indexes.</em></p>
<p>Run
<a href="./Pipeline_B/0b_make_database_and_index.sh">0b_make_database_and_index.sh</a>.</p>
<p>Example usage:</p>
<pre class="hljs"><code><div>bash 0b_make_database_and_index.sh
</div></code></pre>
<p><strong>1) BLAST TE of interest against all available genomes.</strong></p>
<p><em>1a) Use TBLASTN with protein sequence queries.</em></p>
<p>This will identify similar TEs in distantly related species. Output will be nucleotide sequences.</p>
<p>Run
<a href="./Pipeline_B/1a_tblastn_and_extract.sbatch">1a_tblastn_and_extract.sbatch</a>.</p>
<p>Example usage:</p>
<pre class="hljs"><code><div>DIR=test_genome DATABASE=YarrowiaLipolytica_ASM252v1.fa QUERY=L1_ORFp.fasta RESULTSDIR=results sbatch 1a_tblastn_and_extract.sbatch
</div></code></pre>
<p><em>1b) (Optional) Use BLASTN or LASTZ with nucleotide sequence queries.</em></p>
<p>Run
<a href="./Pipeline_B/1b_lastz_and_extract.sbatch">1b_lastz_and_extract.sbatch</a>.</p>
<p>Example usage:</p>
<pre class="hljs"><code><div>GENOMEDIR=test_genome GENOME=YarrowiaLipolytica_ASM252v1.fa QUERYDIR=test_query QUERY=L1_nucl_seqs.fasta RESULTSDIR=results sbatch 1b_lastz_and_extract.sbatch
</div></code></pre>
<p><em>1c) For each genome, combine all identified nucleotide sequences from the previous steps.</em></p>
<p>Run <a href="./Pipeline_B/1c_combine_hits.sbatch">1c_combine_hits.sbatch</a>.</p>
<p>Example usage:</p>
<pre class="hljs"><code><div>SPECIES=YarrowiaLipolytica ELEMENT=L1 LASTZFILE=YarrowiaLipolytica_ASM252v1.fa_L1_nucl_seqs.fasta_lastz.bed TBLASTNFILE=YarrowiaLipolytica_ASM252v1.fa_L1_ORFp.fasta_merged.bed GENOME=YarrowiaLipolytica_ASM252v1.fa RESULTSDIR=results sbatch 1c_combine_hits.sbatch
</div></code></pre>
<p><em>1d) Add header annotations to indicate the genome that each sequence was derived from.</em></p>
<p>Run
<a href="./Pipeline_B/1d_append_name_to_headers.sh">1d_append_name_to_headers.sh</a>.</p>
<p>Example usage:</p>
<pre class="hljs"><code><div>SPECIES=YarrowiaLipolytica ELEMENT=L1 RESULTSDIR=results bash 1d_append_name_to_headers.sh
</div></code></pre>
<p>Repeat screening in an iterative process (e.g. BLAST-ing the new, larger, query dataset against each genome and then combining the output) until no new hits are found.</p>
<p><strong>2) Perform a reciprocal best hit check.</strong></p>
<p><em>2a) Use CENSOR to compare hits against known repeat databases (e.g. RepBase or Dfam).</em></p>
<p>Run
<a href="./Pipeline_B/2a_censor_sequences.sbatch">2a_censor_sequences.sbatch</a>.</p>
<p>Example usage:</p>
<pre class="hljs"><code><div>INDIR=results FILE=YarrowiaLipolytica_L1_combined.fasta OUTDIR=results/censored sbatch 2a_censor_sequences.sbatch
</div></code></pre>
<p><em>2b) Confirm and extract hits that match the correct TE family.</em></p>
<p>Run
<a href="./Pipeline_B/2b_check_censor_output.sbatch">2b_check_censor_output.sbatch</a>.</p>
<p>Example usage:</p>
<pre class="hljs"><code><div>SPECIES=Yarrowia.lipolytica FILE=Yarrowia.lipolytica_L1_combined.fasta GENOME=test_genome/YarrowiaLipolytica_ASM252v1.fa ELEMENT=L1 QUERY=test_query/known_L1_elements_from_repbase.txt CENSORDIR=results/censored sbatch 2b_check_censor_output.sbatch
</div></code></pre>
<p><strong>3) Cluster all sequences obtained from the iterative alignment screening.</strong></p>
<p>Prior to this step, you will need to combine hits from all genomes into one file. Make sure that sequence headers indicate the species that each TE sequence was derived from.</p>
<p>This clustering step is important as it will reveal likely HTT events which are manifested as clusters of highly similar elements that include elements from multiple species. We have found it best to use sequence divergence cut offs that cluster most closely related sequences (e.g. &lt;20% divergent).</p>
<p><em>3a) All-against-all clustering of nucleotide sequences using VSEARCH.</em></p>
<p>You can use full-length nucleotide sequences, or nucleotide sequences of the open reading frames only.</p>
<p>Run
<a href="./Pipeline_B/3a_vsearch_cluster_for_nucleotide_seqs.sbatch">3a_vsearch_cluster_for_nucleotide_seqs.sbatch</a>, changing the clustering identity threshold (ID) as required.</p>
<p>Example usage:</p>
<pre class="hljs"><code><div>INDIR=results/allSpeciesCombined FILE=allSpecies_L1.fasta ID=80 PREFIX=c sbatch 3a_vsearch_cluster_for_nucleotide_seqs.sbatch
</div></code></pre>
<p><em>3b) All-against-all clustering of amino acid sequences using USEARCH.</em></p>
<p>VSEARCH (the open source alternative to USEARCH) does not support protein sequences, but will not fail if given protein sequence input. Make sure you use another program (e.g. Cd-hit or USEARCH) to clustering amino acid sequences. The 32-bit version of USEARCH is open source.</p>
<p>The following script can be used to perform all-against-all cluserting of amino acid sequences from ORFs, or reverse transcriptase domains from retrotransposons/transposase domains from DNA transposons.</p>
<p>Run
<a href="./Pipeline_B/3b_usearch_cluster_for_aa_seqs.sbatch">3b_usearch_cluster_for_aa_seqs.sbatch</a>.</p>
<p>Example usage:</p>
<pre class="hljs"><code><div>INDIR=results/allSpeciesCombined FILE=allSpecies_L1_ORFp.fasta ID=90 PREFIX=r sbatch 3b_usearch_cluster_for_aa_seqs.sbatch
</div></code></pre>
<p><strong>4) Identify clusters containing TE sequences from multiple species (e.g. based on the sequence header names).</strong></p>
<p>These clusters are the HTT candidates.</p>
<p><em><strong>Test dataset for Pipeline B</strong></em></p>
<p>A test genome (fungus <em>Yarrowia lipolytica</em>) has been placed in <a href="./Pipeline_B/test_genome">test_genome</a>, along with a set of L1 repeats as a <a href="./Pipeline_B/test_query">test_query</a>. We recommend trying out the workflow using these files first.</p>
<h3 id="c-htt-candidates-validation-a-name%22workflowc%22">C: HTT candidates validation <a name="workflowC" /></h3>
<p>Pipeline C includes several additional steps to validate detected HTT candidates from Pipeline A or Pipeline B. Manually checking will be required in this section.</p>
<p><em><strong>Usage</strong></em></p>
<p>NOTE: Assume you are in <code>Pipeline_C</code> folder to run these following scripts.</p>
<p><strong>1) Presence/absence analysis based on flanking regions.</strong></p>
<p>Extend individual TE insertions with 1-2kb of flanking sequence and align those sequences to the genomes of related/target species. Use the coordinates from these alignments to find the ends of the target sequences and extract those sequences from the target genomes. The query and target sequences are then globally aligned (MAFFT/MUSCLE) to determine presence/absence compared to ancestral insertions in related taxa (Manually checking in aligment viewer will be required).</p>
<p><strong>2) Generate and cluster species-specific sub-clusters using VSEARCH or USEARCH, and align all sub-clusters to create consensus sequences, or identify centroids.</strong></p>
<p>Run <a href="./Pipeline_C/2_usearchConsensus.sbatch">2_usearchConsensus.sbatch</a></p>
<p>Example usage (please modify job settings according to your HPC environment):</p>
<pre class="hljs"><code><div>INDIR=results FILE=HTT_condidates.fasta ID=80 sbatch 2_usearchConsensus.sbatch
</div></code></pre>
<p><strong>3) Align all consensus sequences/centroids to generate a sequence phylogeny.</strong></p>
<p>Global multisequence alignment of representative sequences to generate a phylogenetic tree of HTT candidate sequences. This phylogenetic distribution of representative sequences can be compared to the species phylogeny to see if it is discordant.</p>
<p>Run
<a href="./Pipeline_C/3_alignRefineAndTree.sbatch">3_alignRefineAndTree.sbatch</a></p>
<p>Example usage (please modify job settings according to your HPC
environment):</p>
<pre class="hljs"><code><div>INDIR=results FILE=HTT_condidates_centroids.fasta MINBLOCKSIZE=5 ALLOWEDGAPS=a sbatch 3_alignRefineAndTree.sbatch
</div></code></pre>
<p><strong>4) Further analysis</strong></p>
<p>Divergence analysis of HTT candidate clusters to determine activity profile and ensure that this profile is consistent with transfer from the most closely related TE from another species. Additional/alternative methods might include:</p>
<ul>
<li>
<p>using a k-mer based method (e.g Jellyfish) to compare k-mers within TE sequences. Jellyfish provides an alignment-free alternative for detecting similarities in TEs across species.</p>
</li>
<li>
<p>using intact ORFs to determine how recently TEs in different species were likely active, to estimate the timing of potential HTT events.</p>
</li>
<li>
<p>BLAST-ing sequences from HTT candidate clusters against all eukaryotes on NCBI (not just your database of genomes) to identify potential vector/source species.</p>
</li>
</ul>
<h2 id="additional-notes-a-name%22notes%22">Additional notes <a name="notes" /></h2>
<ul>
<li>
<ol>
<li>Software based TE annotation must be manually curated prior to
analysis, as no software tool is guaranteed to find full length
TEs, or find all full length TEs.</li>
</ol>
</li>
<li>
<ol start="2">
<li>For TEs with two open reading frames, concatenating ORF1 and
ORF2 sequences before running TBLASTN may give the best results.</li>
</ol>
</li>
<li>
<ol start="3">
<li>We recommend using a combination of TBLASTN (protein sequence
input, e.g. TE ORFs) and BLASTN (nucleotide sequence input,
e.g. full-length TE) for identifying horizontally transferred
TEs in a wide range of genomes. If only using one method,
TBLASTN should be prioritised, especially when looking for TEs
of the same family in distantly related species. Nucleotide
queries (e.g. with BLASTN) may fail to identify TEs that have
had longer to mutate/diverge.</li>
</ol>
</li>
<li>
<ol start="4">
<li>VSEARCH (the open source alternative to USEARCH) does not
support protein sequences, but will not fail if given protein
sequence input. Make sure to use another program (e.g. Cd-hit or
USEARCH) for clustering of amino acid sequences. The 32-bit
version of USEARCH is open source.</li>
</ol>
</li>
<li>
<ol start="5">
<li>Confirmation of presence/absence is only practical for small
numbers of HTT events that can be evaluated by eye.</li>
</ol>
</li>
<li>
<ol start="6">
<li>Alignment viewers such as JalView can also perform principal
component analysis (PCA) of all sequences in a multiple sequence
alignment, where each dot represents an individual TE sequence,
and dots can be coloured by species. This allows easy
visualisation of HTT clusters containing highly similar
sequences from different species.</li>
</ol>
</li>
</ul>
<h2 id="references">References</h2>
<ul>
<li>Galbraith JD, Ludington AJ, Suh A, Sanders KL, Adelson DL. New Environment, New Invaders-Repeated Horizontal Transfer of LINEs to Sea Snakes. <em>Genome Biol Evol.</em> 2020;12:2370--83.</li>
<li>Galbraith JD, Ludington AJ, Sanders KL, Suh A, Adelson DL. Horizontal transfer and subsequent explosive expansion of a DNA transposon in sea kraits (Laticauda). <em>Biol Lett.</em> 2021;17:20210342.</li>
<li>Ivancevic AM, Kortschak RD, Bertozzi T, Adelson DL. Horizontal transfer of BovB and L1 retrotransposons in eukaryotes. <em>Genome Biol.</em> 2018;19:85.</li>
<li>Ivancevic, Atma M, Kortschak, R Daniel, Bertozzi, Terry, &amp; Adelson, David L. (2018). Dataset from: Horizontal transfer of BovB and L1 retrotransposons in eukaryotes [Data set]. <em>Zenodo.</em> <a href="http://doi.org/10.5281/zenodo.1246946">http://doi.org/10.5281/zenodo.1246946</a></li>
<li>Ivancevic Atma M. (2018, May 15). AdelaideBioinfo/horizontalTransfer: First release of horizontal transfer code (Version v1.0.0). <em>Zenodo.</em> <a href="http://doi.org/10.5281/zenodo.1246999">http://doi.org/10.5281/zenodo.1246999</a></li>
</ul>
<h2 id="the-end-a-name%22the-end%22">The end <a name="the-end" /></h2>
<p><em>“Remember that all models are wrong; the practical question is how wrong do they have to be to not be useful.”</em> George Box in; Box, G.E.P. and Draper, N.R. Empirical Model Building and Response Surfaces. John Wiley &amp; Sons, New York. (1987).</p>

</body>
</html>
